### PR TITLE
dev/core#1093 add a bulkCreate action for many customFields in one go

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
Add bulkCreate function for CustomField with a view towards this being a new protocol for how bulk create actions would look from a code POV which we could expose via apiv4

Before
----------------------------------------
No function for efficiently creating multiple fields

After
----------------------------------------
Function exists. There is a view to expose this via apiv4 but at this stage that has been left off to allow 'maturing' - unit test added

Technical Details
----------------------------------------
Preliminary cleanup changes are split off into smaller commits 

https://github.com/civicrm/civicrm-core/pull/14705
https://github.com/civicrm/civicrm-core/pull/14697
https://github.com/civicrm/civicrm-core/pull/14693 (merged)
https://github.com/civicrm/civicrm-core/pull/14689 (merged)

Comments
----------------------------------------

I haven't switched Utils_Migrate to use the bulk action but that is the next logical step since it is kinda creating complexity to do 'bulk-action-lite'

https://lab.civicrm.org/dev/core/issues/1093

### CODE REVIEW

**Comment by pradpnayak** [comment]
This needs rebasing

**Comment by eileenmcnaughton** [comment]
rebased

**Comment by colemanw** [comment]
Api4 signature could look like this: https://github.com/civicrm/org.civicrm.api4/issues/180

**Comment by colemanw** [comment]
@eileenmcnaughton looks like a brittle test fell over on a whitespace change made in this PR:

    CRM_Core_BAO_SchemaHandlerTest::testBuildFieldChangeSql
    Failed asserting that two strings are equal.
    --- Expected
    +++ Actual
    @@ @@
    -'ALTER TABLE big_table\n
    +'ALTER TABLE big_table \n

**Comment by eileenmcnaughton** [comment]
@colemanw good spotting - that was also what the sub-pr got stuck on - https://github.com/civicrm/civicrm-core/pull/14727

**Comment by eileenmcnaughton** [comment]
@colemanw now all the preliminary refactoring is done this is pretty simple - I have renamed it to bulkSave per your recommendation

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by eileenmcnaughton** [comment]
@colemanw we are down to one unrelated fail here!

File: CRM/Core/BAO/CustomField.php [hunk 1]
```diff
@@ -1650,9 +1685,47 @@ public static function defaultCustomTableSchema($params) {
    * @param bool $triggerRebuild
    */
   public static function createField($field, $operation, $indexExist = FALSE, $triggerRebuild = TRUE) {
+    $params = [
+      'table_name' => CRM_Core_DAO::getFieldValue('CRM_Core_DAO_CustomGroup', $field->custom_group_id, 'table_name'),
+    ];
+    $sql = str_repeat(' ', 8);
```
Comment (colemanw) [inline CRM/Core/BAO/CustomField.php]: ????????

File: CRM/Core/BAO/CustomField.php [hunk 2]
```diff
@@ -164,6 +164,41 @@ public static function create($params) {
...
+   *  Default parameters to be be merged into each of the params.
+   */
+  public static function bulkSave($bulkParams, $defaults) {
+    $sql = [];
+    $customFields = [];
+    foreach ($bulkParams as $fieldParams) {
+      $params = array_merge($defaults, $fieldParams);
+      $operation = empty($params['id']) ? 'add' : 'modify';
+      $customField = self::createCustomFieldRecord($params);
+      // Only doing 'add' for now - so hard coded & no indexExists.
```
Comment (colemanw) [inline CRM/Core/BAO/CustomField.php]: What would it take to get this working with update? Then it would be ready for api4.

Comment (eileenmcnaughton) [reply CRM/Core/BAO/CustomField.php]: oh - I think I mostly altered it but not in this spot - will fix

File: CRM/Core/BAO/CustomField.php (hunk 1 above)
Comment (eileenmcnaughton) [reply CRM/Core/BAO/CustomField.php]: yeah weird but retained from 'create'

**Comment by eileenmcnaughton** [comment]
@colemanw it passed! (& test cover was good - I couldn't sneak anything past it)

File: CRM/Core/BAO/SchemaHandler.php [hunk 3]
```diff
@@ -295,7 +295,7 @@ public static function buildForeignKeySQL($params, $separator, $prefix, $tableNa
    * @return bool
    */
   public static function alterFieldSQL($params, $indexExist = FALSE, $triggerRebuild = TRUE) {
-
+    CRM_Core_Error::deprecatedFunctionWarning('function no longer in use / supported');
```
Comment (colemanw) [inline CRM/Core/BAO/SchemaHandler.php]: Probably should add a `@deprecated` annotation too.

Comment (eileenmcnaughton) [reply CRM/Core/BAO/SchemaHandler.php]: Ok - I'll add that to the last PR (rather than invalidate tests on this)

**Comment by colemanw** [comment]
Code & test coverage looks good. Merging.

**Pull request merged**